### PR TITLE
gui: Add better support for high DPI displays

### PR DESCRIPTION
**Pull request by Yunotchi**
# About
This PR is carrying on work from #1220.
- Vita3K's window size will scale depending on the user's OS scaling setting -- currently only implemented for Windows
- Fonts will render at a higher resolution to match the larger window size
- Back button's appearance is more consistent across system apps
- Images (e.g. icons and background images) scale nicer
  - Switched to linear filtering from nearest neighbour. This could have knock-on effects that I am not aware of; please check my change to `vita3k\gui\src\imgui_impl_sdl_gl3.cpp`
- App selector's icons now scale with window size
- Content manager: black bar at bottom now fills the whole window's width

# Screenshots
Open in separate tabs to compare before and after.

| Before | After |
| --- | --- |
|  |  |
|  |  |
|  |  |
|  |  |

# Ideas
- Initial icon size could vary based on dpi setting
- Dynamically adjust font resolution when resizing window

# Known issues
- ~Some buttons and icons are slightly mispositioned~ Should all be fixed
- ~Settings app: value picker for Date & Time, Language - tick mark looks wrong~ Fixed
- ~Item/frame/window rounding (notable on lock screen) not scaled correctly~ Should all be fixed

### CODE REVIEW

File: vita3k/gui/src/imgui_impl_sdl_gl3.cpp [hunk 1]
```diff
@@ -406,8 +406,8 @@ IMGUI_API ImTextureID ImGui_ImplSdlGL3_CreateTexture(void *data, int width, int
     glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
     glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
     glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_R, GL_REPEAT);
-    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
```
Comment (Yunotchi) [inline vita3k/gui/src/imgui_impl_sdl_gl3.cpp:410]: Could this cause problems or undesired effects?

**Comment by Zangetsu38** [comment]
Squash, no create new commit but edit last and force push

**Comment by Zangetsu38** [comment]
little fail, remove change relate menu_ bar for background, i have set t'his for good reason, here you literally cut and deformed bg lol

File: vita3k/gui/src/content_manager.cpp [hunk 2]
```diff
@@ -255,9 +255,9 @@ void draw_content_manager(GuiState &gui, HostState &host) {
 
     ImGui::Begin("##content_manager", &gui.live_area.content_manager, ImGuiWindowFlags_NoDecoration | ImGuiWindowFlags_NoSavedSettings);
     if (is_background)
-        ImGui::GetBackgroundDrawList()->AddImage(gui.apps_background["NPXS10026"], ImVec2(0.f, MENUBAR_HEIGHT), display_size);
+        ImGui::GetBackgroundDrawList()->AddImage(gui.apps_background["NPXS10026"], ImVec2(0.f, 0.f), display_size);
```
Comment (Zangetsu38) [inline vita3k/gui/src/content_manager.cpp:258]: revert this

File: vita3k/gui/src/live_area.cpp [hunk 3]
```diff
@@ -817,7 +817,7 @@ void draw_live_area_screen(GuiState &gui, HostState &host) {
 
     if (is_background)
         ImGui::GetBackgroundDrawList()->AddImage((gui.users[host.io.user_id].use_theme_bg && !gui.theme_backgrounds.empty()) ? gui.theme_backgrounds[gui.current_theme_bg] : gui.user_backgrounds[gui.users[host.io.user_id].backgrounds[gui.current_user_bg]],
-            ImVec2(0.f, 32.f), display_size);
+            ImVec2(0.f, 0.f), display_size);
```
Comment (Zangetsu38) [inline vita3k/gui/src/live_area.cpp:820]: keep 32, image obligatoryè start after information bar

File: vita3k/gui/src/settings.cpp [hunk 4]
```diff
@@ -500,9 +500,9 @@ void draw_start_screen(GuiState &gui, HostState &host) {
     ImGui::Begin("##start_screen", &gui.live_area.start_screen, ImGuiWindowFlags_NoBackground | ImGuiWindowFlags_NoDecoration | ImGuiWindowFlags_NoSavedSettings);
 
     if (gui.start_background)
-        ImGui::GetForegroundDrawList()->AddImage(gui.start_background, ImVec2(0.f, MENUBAR_HEIGHT), display_size);
+        ImGui::GetBackgroundDrawList()->AddImage(gui.start_background, ImVec2(0.f, 0.f), display_size);
```
Comment (Zangetsu38) [inline vita3k/gui/src/settings.cpp:503]: rever

File: vita3k/gui/src/settings.cpp [hunk 5]
```diff
@@ -500,9 +500,9 @@ void draw_start_screen(GuiState &gui, HostState &host) {
...
     else
-        ImGui::GetForegroundDrawList()->AddRectFilled(ImVec2(0.f, MENUBAR_HEIGHT), display_size, IM_COL32(128.f, 128.f, 128.f, 128.f), 0.f, ImDrawCornerFlags_All);
+        ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, 0.f), display_size, IM_COL32(128.f, 128.f, 128.f, 128.f), 0.f, ImDrawCornerFlags_All);
```
Comment (Zangetsu38) [inline vita3k/gui/src/settings.cpp:505]: same

File: vita3k/gui/src/trophy_collection.cpp [hunk 6]
```diff
@@ -414,9 +414,9 @@ void draw_trophy_collection(GuiState &gui, HostState &host) {
     ImGui::PushStyleVar(ImGuiStyleVar_WindowBorderSize, 0.f);
     ImGui::Begin("##trophy_collection", &gui.live_area.trophy_collection, ImGuiWindowFlags_NoDecoration | ImGuiWindowFlags_NoSavedSettings);
     if (is_background)
-        ImGui::GetBackgroundDrawList()->AddImage(gui.apps_background["NPXS10008"], ImVec2(0.f, MENUBAR_HEIGHT), display_size);
+        ImGui::GetBackgroundDrawList()->AddImage(gui.apps_background["NPXS10008"], ImVec2(0.f, 0.f), display_size);
```
Comment (Zangetsu38) [inline vita3k/gui/src/trophy_collection.cpp:417]: rever

File: vita3k/gui/src/user_management.cpp [hunk 7]
```diff
@@ -237,12 +237,12 @@ void draw_user_management(GuiState &gui, HostState &host) {
     if (!host.display.imgui_render || ImGui::IsWindowHovered(ImGuiHoveredFlags_RootAndChildWindows))
         gui.live_area.information_bar = true;
 
-    ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, MENUBAR_HEIGHT), display_size, IM_COL32(10.f, 50.f, 140.f, 255.f), 0.f, ImDrawCornerFlags_All);
+    ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, 0.f), display_size, IM_COL32(10.f, 50.f, 140.f, 255.f), 0.f, ImDrawCornerFlags_All);
```
Comment (Zangetsu38) [inline vita3k/gui/src/user_management.cpp:240]: rever

File: vita3k/gui/src/app_selector.cpp [hunk 8]
```diff
@@ -240,21 +240,21 @@ void draw_app_selector(GuiState &gui, HostState &host) {
 
     if (is_background)
         ImGui::GetBackgroundDrawList()->AddImage((gui.users[host.io.user_id].use_theme_bg && !gui.theme_backgrounds.empty()) ? gui.theme_backgrounds[gui.current_theme_bg] : gui.user_backgrounds[gui.users[host.io.user_id].backgrounds[gui.current_user_bg]],
-            ImVec2(0.f, MENUBAR_BG_HEIGHT), display_size);
+            ImVec2(0.f, 0.f), display_size);
```
Comment (Zangetsu38) [inline vita3k/gui/src/app_selector.cpp:243]: revert

File: vita3k/gui/src/app_selector.cpp [hunk 9]
```diff
@@ -240,21 +240,21 @@ void draw_app_selector(GuiState &gui, HostState &host) {
...
     else
-        ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, MENUBAR_BG_HEIGHT), display_size, IM_COL32(11.f, 90.f, 252.f, 160.f), 0.f, ImDrawCornerFlags_All);
+        ImGui::GetBackgroundDrawList()->AddRectFilled(ImVec2(0.f, 0.f), display_size, IM_COL32(11.f, 90.f, 252.f, 160.f), 0.f, ImDrawCornerFlags_All);
```
Comment (Zangetsu38) [inline vita3k/gui/src/app_selector.cpp:245]: same

File: vita3k/gui/src/controls_dialog.cpp [hunk 10]
```diff
@@ -61,11 +61,13 @@ static void remapper_button(GuiState &gui, HostState &host, int *button, const c
...
 void draw_controls_dialog(GuiState &gui, HostState &host) {
-    float width = ImGui::GetWindowWidth() / 1.35f;
-    float height = ImGui::GetWindowHeight() / 1.25f;
+    float width = ImGui::GetWindowWidth() / 1.35f * gui.dpi_scale;
+    float height = ImGui::GetWindowHeight() / 1.25f * gui.dpi_scale;
     ImGui::SetNextWindowSize(ImVec2(width, height));
     ImGui::SetNextWindowPos(ImVec2(ImGui::GetIO().DisplaySize.x / 2.f, ImGui::GetIO().DisplaySize.y / 2.f), ImGuiCond_Always, ImVec2(0.5f, 0.5f));
     ImGui::Begin("Controls", &gui.controls_menu.controls_dialog);
+    //ImGui::SetNextWindowPos(ImVec2(ImGui::GetIO().DisplaySize.x / 2.f, ImGui::GetIO().DisplaySize.y / 2.f), ImGuiCond_Always, ImVec2(0.5f, 0.5f));
+    //ImGui::Begin("Controls", &gui.controls_menu.controls_dialog, ImGuiWindowFlags_AlwaysAutoResize);
```
Comment (Zangetsu38) [inline vita3k/gui/src/controls_dialog.cpp:70]: 🤔  ?

File: vita3k/gui/src/imgui_impl_sdl_gl3.cpp (hunk 1 above)
Comment (Zangetsu38) [reply vita3k/gui/src/imgui_impl_sdl_gl3.cpp:410]: ask @1whatleytay

File: vita3k/gui/src/controls_dialog.cpp (hunk 10 above)
Comment (Yunotchi) [reply vita3k/gui/src/controls_dialog.cpp:70]: I will fix this lol

**Comment by Yunotchi** [comment]
> psvita theme bg is 960x512, because is 544- Information bar (32)

The background images in `AppData\Roaming\Vita3K\Vita3K\vs0\app` (e.g. `\NPXS10002\sce_sys\pic0.png`) are all 960x544, and official wallpapers provided by game companies are 960x544 (e.g. https://aquaplus.jp/dt2vita/wallpaper/) 

Are the themes that you buy from PSN different?

**Comment by Zangetsu38** [comment]
> 
> 
> > psvita theme bg is 960x512, because is 544- Information bar (32)
> 
> The background images in `AppData\Roaming\Vita3K\Vita3K\vs0\app` (e.g. `\NPXS10002\sce_sys\pic0.png`) are all 960x544, and official wallpapers provided by game companies are 960x544 (e.g. https://aquaplus.jp/dt2vita/wallpaper/)
> 
> Are the themes that you buy from PSN different?

ah yes sorry, i forget i u_sing app bg for system app lol, is like splash screen, so yes is 544 but, for theme is 512, so for start screen and app selector keep menubar

**Comment by Zangetsu38** [comment]
also on psvita, if you using personal bg for start screen on wallpaper, it obligatory resize image for start after information bar, for no cut image and no deforme, is for that on psvita, it crop image

**Comment by Yunotchi** [comment]
Ah, thank you for explaining
I'll take another look at this soon :)

**Comment by Zangetsu38** [comment]
> 
> 
> Ah, thank you for explaining
> I'll take another look at this soon :)

i have remove my comment on change so no request,  so just loop where is keep ^^

File: vita3k/app/src/app_init.cpp [hunk 11]
```diff
@@ -151,7 +151,10 @@ bool init(HostState &state, Config &cfg, const Root &root_paths, CPUDepInject in
     }
 #ifdef WIN32
     SetProcessDpiAwareness(PROCESS_PER_MONITOR_DPI_AWARE);
-    state.window = WindowPtr(SDL_CreateWindow(window_title, SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, DEFAULT_RES_WIDTH, DEFAULT_RES_HEIGHT, window_type | SDL_WINDOW_RESIZABLE | SDL_WINDOW_ALLOW_HIGHDPI), SDL_DestroyWindow);
+    float ddpi, hdpi, vdpi;
+    SDL_GetDisplayDPI(0, &ddpi, &hdpi, &vdpi);
+    float dpi_scale = ddpi / 96;
```
Comment (Zangetsu38) [inline vita3k/app/src/app_init.cpp:156]: try keep this value here for no repeat this code in other part, with adding GuiState or with store value inside state (host)

File: vita3k/WindowsDpiAwareness.manifest [hunk 12]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8" standalone="yes"?>
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0" xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
+  <asmv3:application>
+    <asmv3:windowsSettings>
+      <dpiAware xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">true</dpiAware>
+      <dpiAwareness xmlns="http://schemas.microsoft.com/SMI/2016/WindowsSettings">PerMonitorV2</dpiAwareness>
+    </asmv3:windowsSettings>
+  </asmv3:application>
+</assembly>
```
Comment (Zangetsu38) [inline vita3k/WindowsDpiAwareness.manifest:9]: never forget add new line in end file

File: vita3k/gui/src/app_selector.cpp [hunk 13]
```diff
@@ -184,13 +184,13 @@ static float scroll_type, current_scroll_pos, max_scroll_pos;
 void draw_app_selector(GuiState &gui, HostState &host) {
     const ImVec2 display_size = ImGui::GetIO().DisplaySize;
     const auto scal = ImVec2(display_size.x / 960.0f, display_size.y / 544.0f);
-    const auto MENUBAR_HEIGHT = 32.f * scal.y;
+    const auto INFORMATION_BAR_HEIGHT = 32.f * scal.y;
     const auto MENUBAR_BG_HEIGHT = !gui.live_area.information_bar ? 22.f : 32.f * scal.y;
```
Comment (Zangetsu38) [inline vita3k/gui/src/app_selector.cpp:188]: AH my bad lol if yçou can change this to 
```suggestion
    const auto MENUBAR_BG_HEIGHT = !gui.live_area.information_bar ? 22.f : INFORMATION_BAR_HEIGHT ;
```
thx

**Comment by Yunotchi** [comment]
Made a few changes:
- DPI awareness is done in cmake with manifest file -- this is better than doing it at runtime, and now the application loads a higher resolution version of its icon
- Reverted changes to backgrounds -- I want to look at this again later
- Notification icon replaced with more authentic circle
- Center user names in User Management
- DPI scaling is done with `ImGuiIO.DisplayFramebufferScale`. Fonts are loaded at a larger size and scaled down globally via `ImGuiIO.FontGlobalScale`
- Renamed `MENUBAR_HEIGHT` to `INFORMATION_BAR_HEIGHT` in the gui classes

File: vita3k/gui/src/information_bar.cpp [hunk 14]
```diff
@@ -356,7 +356,7 @@ static void draw_notice_info(GuiState &gui, HostState &host) {
         if (gui.theme_information_bar_notice.find("no") != gui.theme_information_bar_notice.end())
             ImGui::GetForegroundDrawList()->AddImage(gui.theme_information_bar_notice["no"], NOTICE_POS, ImVec2(NOTICE_POS.x + NOTICE_SIZE.x, NOTICE_SIZE.y));
         else
-            ImGui::GetForegroundDrawList()->AddRectFilled(ImVec2(display_size.x - (70.f * SCAL.x), (-24.f * SCAL.y)), ImVec2(display_size.x + (14.f * SCAL.y), 60.f * SCAL.y), IM_COL32(62.f, 98.f, 160.f, 255.f), 75.f * SCAL.x, ImDrawCornerFlags_All);
+            ImGui::GetForegroundDrawList()->AddCircleFilled(ImVec2(display_size.x - (28.f * SCAL.x), (18.f * SCAL.x)), 44.0f * SCAL.x, IM_COL32(62.f, 98.f, 160.f, 255.f));
```
Comment (Zangetsu38) [inline vita3k/gui/src/information_bar.cpp:359]: i have try is great :D, add same on 353 line just before, is notif new, size is little more big try calculate for get same position :)

**Comment by Yunotchi** [comment]
Made some more changes, hopefully final for this PR:
- Added scaling for save dialog windows
- Fixed button rounding in many places not scaling correctly with window size
- Reverted changes to notification icon -- I will do this and other GUI tweaks in another PR
- Live area: tweaked start button gate's rounding and thickness to look more authentic
- App selector: icons now scale with window size
- Content manager: black bar at bottom now fills the whole window's width

File: vita3k/gui/src/gui.cpp [hunk 15]
```diff
@@ -105,7 +107,16 @@ static void init_style() {
 
 static void init_font(GuiState &gui, HostState &host) {
     ImGuiIO &io = ImGui::GetIO();
+
+#ifdef _WIN32
+    if (host.dpi_scale > 1) {
+        gui.monospaced_font = io.Fonts->AddFontFromFileTTF("C:\\Windows\\Fonts\\consola.ttf", 13.0f * host.dpi_scale * host.dpi_scale, NULL, io.Fonts->GetGlyphRangesJapanese()); // multiply by dpi_scale twice to correctly scale
```
Comment (pent0) [inline vita3k/gui/src/gui.cpp:113]: Constant string should be made a separate variable with name explaining the usage

File: vita3k/gui/src/information_bar.cpp-d13cf8a6 [hunk 16]
```diff
@@ -0,0 +1,562 @@
+﻿// Vita3K emulator project
```
Comment (pent0) [inline vita3k/gui/src/information_bar.cpp-d13cf8a6:1]: This file is unintentionally left over?

**Review by pent0 — CHANGES_REQUESTED** [review]
I can't judge much about how you implement it, there are still some more nits

**Comment by Yunotchi** [comment]
Thank you for the feedback, pent0. I've addressed the issues you raised.

I have also reverted the implementation of DPI-awareness I made in #1220 because:
- shcore isn't available in Windows 7
- as I mentioned in an above comment, implementing DPI-awareness via an application manifest is better

**Comment by scribam** [comment]
I am not sure if it can be useful but there is a function in ImGui about DPI awareness : https://github.com/ocornut/imgui/blob/7b8932554e04aca62d3c2d4ffbb5b9212a55f801/backends/imgui_impl_win32.cpp#L444

**Comment by Yunotchi** [comment]
Hm, I'll look at that ImGui function to see what it does.

My current implementation is working correctly on Windows 10 -- see the last column in this screenshot:
image

I don't have a Windows 7 machine to test with but [Microsoft's documentation](https://docs.microsoft.com/en-us/windows/win32/hidpi/setting-the-default-dpi-awareness-for-a-process) says it should work on all Windows since Vista.

**Comment by sunho** [comment]
Can you rebase this pr.

File: vita3k/gui/src/common_dialog.cpp [hunk 17]
```diff
@@ -60,18 +61,22 @@ static void draw_ime_dialog(DialogState &common_dialog) {
 }
 
 static void draw_message_dialog(DialogState &common_dialog) {
-    const ImVec2 BUTTON_SIZE = ImVec2(200.f, 35.f);
-    const ImVec2 WINDOW_SIZE = ImVec2(ImGui::GetIO().DisplaySize.x / 1.7f, ImGui::GetIO().DisplaySize.y / 1.5f);
-    const ImVec2 PROGRESS_BAR_SIZE = ImVec2(WINDOW_SIZE.x - 120.f, 7.f);
+    LOG_INFO("draw_message_dialog");
```
Comment (Zangetsu38) [inline vita3k/gui/src/common_dialog.cpp:64]: Hmmmmm ?

Comment (Yunotchi) [reply vita3k/gui/src/common_dialog.cpp:64]: Darn it, will fix

**Review by pent0 — DISMISSED** [review]
About the Linear change, it was for font blurry I think, but if you do DPI scale I don't think it should be a worry to not use Nearest anymore. But take notes if anything degrades (doubts).

Edit: Rebase then approve again.

**Review by pent0 — APPROVED** [review]

File: vita3k/gui/src/common_dialog.cpp (hunk 17 above)
Comment (Yunotchi) [reply vita3k/gui/src/common_dialog.cpp]: Removed

**Comment by Yunotchi** [comment]
OK, I've rebased my branch.

> About the Linear change, it was for font blurry I think, but if you do DPI scale I don't think it should be a worry to not use Nearest anymore. But take notes if anything degrades (doubts).

My reason for changing from nearest neighbour to linear was for icons:
1 5DPI-windowed-2x

**Pull request merged**